### ISSUE TITLE
Bug Report:  Missing example values in OpenApi components schema definition using nested objects - FISH-1204

### ISSUE THREAD

**Issue by LarsBrenker** [payara/Payara#5151]
# Description #
----------

When using nested objects, it was noticed that there is no example value on the nested object parameters. 
As a result, there is no example value in the request body example.

## Expected Outcome ##

![grafik](https://user-images.githubusercontent.com/57911568/110018183-551b9680-7d27-11eb-967d-565d33511168.png)

## Current Outcome ##

![grafik](https://user-images.githubusercontent.com/57911568/110018334-78dedc80-7d27-11eb-85f4-431d1381b3e3.png)

The problem is that the example values ​​in the schema of the nested object "ShipmentData" are not generated on the "Partner" object :

(For the sake of clarity, I've removed some attributes)
![grafik](https://user-images.githubusercontent.com/57911568/110019210-7af56b00-7d28-11eb-866a-59de61c9d6aa.png)

I think there are two ways to solve the problem:

1. If the object contains the nested object, the example values ​​must also be adopted
![grafik](https://user-images.githubusercontent.com/57911568/110020039-65347580-7d29-11eb-8300-d5a0eb66ecca.png)

2. Link the nested object to the object by reference
![grafik](https://user-images.githubusercontent.com/57911568/110020160-89905200-7d29-11eb-9590-17bd74ee60f8.png)

The second solution looks nicer to me personally because it makes the document clearer.

## Steps to reproduce ## 

Reproducer: https://github.com/LarsBrenker/pay5151

## Environment ##

- **Distribution**: Micro
- **JDK Version**: 11 Open JDK
- **Operating System**: Official Payara Docker Image 5.2021.1 JDK 11

**Comment by MeroRai** [comment]
I have could reproduce the issue in our current release of Payara community release. I have now escalated it to our engineering team and created an internal issue `FISH-1204` to track it. Thank you for your assist.

**Comment by aubi** [comment]
Hello @LarsBrenker ,
I'm working on this issue. I agree, that reference is a better way, but in this case, the description provided on the reference will be lost:
```java
   @Schema(required = true, name = "shipmentData", description = "THIS WOULD BE LOST", implementation = ShipmentData.class)
    private ShipmentData shipmentData;
```

So for this case, in order to keep all information, we will make the copy including the examples.
But in case there is no other information in schema, we will make a reference:
```java
   @Schema(required = true, implementation = ShipmentData.class)
    private ShipmentData shipmentData;
```
The "required" attribute can be specified as it is recorded in the parent schema.

Because this behaviour can be mysterious (sometimes we generate ref and sometimes not), I plan to add a warning into log, when we cannot make a reference.

Does this make sense to you?
Petr

**Comment by LarsBrenker** [comment]
Hi @aubi ,

if I understood correctly, siblings elements are ignored when a reference is used, right?
So either reference or copied values?

However, I do not yet understand what you are proposing. In this case you would copy the values ​​correctly but issue a warning in the log that the use of reference is not possible due to the additional parameters?

**Comment by LarsBrenker** [comment]
Hi @aubi ,

I've talked to a few developers and we think it's not a good idea to implement two ways. From the developer's point of view, a uniform approach would be preferable. In this case, always copy the values ​​and not use the reference.

A two-pronged implementation only increases the complexity unnecessarily and is again more prone to errors.

The only disadvantage would be that the OpenApi document can be large. From our point of view, that doesn't matter. Developers who take the Code First approach do not look at the document anyway. Most of them look in OpenApi UI. The developers who choose Open API first create the document manually anyway.

So if you ask us: A uniform way and always copy the values.

Especially since this log has no effect anyway. The developer will ignore the log, because if he really wants to use the description, he doesn't care whether a log appears or not.

**Comment by aubi** [comment]
@LarsBrenker 
ad sibling elements: yes, this is the situation. Either ref or anything else. The only exception is "required" as it is mentioned in OpenApi in the parent schema:
```java
    @Schema(required = true, name = "shipmentData", implementation = ShipmentData.class)
    private ShipmentData shipmentData;
```
```yaml
components:
  schemas:
    Partner:
      required:
      - shipmentData
      type: object
      properties:
        shipmentData:
```

**Comment by aubi** [comment]
@LarsBrenker 
https://javadoc.io/static/org.eclipse.microprofile.openapi/microprofile-openapi-api/2.0/org/eclipse/microprofile/openapi/annotations/media/Schema.html#ref--
ref: "This property and all other properties are mutually exclusive."

OK, my original proposal can be confusing. Actually, there is a way to generate a reference (verified):
```java
    @Schema(ref = "#/components/schemas/ShipmentData")
    private ShipmentData shipmentData;
```

I was verifying your project on the current *payara-micro-5.2021.4.jar* and it seems, that the copying of the values works well:

```yaml
components:
  schemas:
    Partner:
      properties:
        shipmentData:
          exclusiveMaximum: false
          exclusiveMinimum: false
          minLength: 0
          uniqueItems: false
          maxProperties: 0
          minProperties: 0
          type: object
          properties:
            salutation:
              type: string
              description: Salutation of the delivery address contact person
              example: MR
              enum:
              - MR
              - MS
            quantity:
              type: integer
              description: Item quantity
              example: "1"
```

I also tried the openapi-ui:
```yaml
{
  "id": "1",
  "legal": "GmbH",
  "shipmentData": {
    "salutation": "MR",
    "quantity": "1"
  }
}
```

**Comment by LarsBrenker** [comment]
Hi @aubi , thanks for your feedback.

I looked at that with the ref specification in the schema annotation. You are right, the example values ​​are copied. But unfortunately the "required" specification is missing again. So no real alternative either.

![grafik](https://user-images.githubusercontent.com/57911568/123065080-20833600-d40f-11eb-8556-a121cfd327c3.png)

But I have now understood you correctly that you provide a bugfix that copies the values ​​to the request body when you work with the implementation attribute, right?

**Comment by aubi** [comment]
@LarsBrenker 
Lars, do you use Payara 5.2021.4 - the current one? In this version, the required attribute is propagated correctly:
`java -jar payara-micro-5.2021.4.jar --deploy pay5151/target/payara-5151-reproducer.war`
![image](https://user-images.githubusercontent.com/4932071/123075323-6c86a880-d418-11eb-99f3-249b1900116d.png)

I gave up making the change :-) Copying works well, required attribute works well and if anybody needs reference, he should use `ref`.

**Comment by aubi** [comment]
@LarsBrenker 
Ah, sorry, I see it! You are talking about
```
    @Schema(required = true, ref = "#/components/schemas/ShipmentData")
    private ShipmentData shipmentData1;
```
The combination of required and ref really doesn't work. I'm looking into it.
Thanks for your help!
Petr

**Comment by LarsBrenker** [comment]
Sorry, not that we're talking past each other now.

If I use ref, everything is generated correctly.

```java    
@Schema(required = true,  ref = "#/components/schemas/ShipmentData")
private ShipmentData shipmentData;
```

![grafik](https://user-images.githubusercontent.com/57911568/123108567-d021cd80-d43a-11eb-9be5-2bad6d6f91d8.png)

![grafik](https://user-images.githubusercontent.com/57911568/123108641-e2037080-d43a-11eb-8946-e7ef7a4fccf5.png)

If I use implementation, the sample values ​​of the nested object "ShipmentData" are not displayed in the request area. Only in Schema Section. When using implementation, the values ​​are not copied correctly!

```java    
@Schema(required = true,  implementation = ShipmentData.class)
private ShipmentData shipmentData;
```
![grafik](https://user-images.githubusercontent.com/57911568/123107919-407c1f00-d43a-11eb-88dc-39a1c5315f9f.png)

![grafik](https://user-images.githubusercontent.com/57911568/123107989-4f62d180-d43a-11eb-8475-5044815b86e1.png)

So please don't give up making a change :) We need a bugfix if someone like us works with the implementation attribute

**Comment by aubi** [comment]
Hi @LarsBrenker ,
can you verify, that the output is exactly as required? Up to my knowledge it's ok:
[openapi.yaml.zip](https://github.com/payara/Payara/files/6749209/openapi.yaml.zip)

I added one more test to your project, in class Partner I added a required reference:
```
@Schema(required = true, ref = "#/components/schemas/ShipmentData")
private ShipmentData shipmentDataRef;
```

**Comment by LarsBrenker** [comment]
@aubi Hi Petr, I just looked at it in the openapi-ui and I think it fits. Thank you for your support.

**Comment by aubi** [comment]
@LarsBrenker 
Hi Lars, the change is merged and will be in Payara 5.2021.5.
Thank you for your report and help, I appreciate it!
Petr

**Comment by LarsBrenker** [comment]
Can be closed. Has already been corrected.
